### PR TITLE
home page

### PR DESCRIPTION
**Pull request by kiahjh**

### CODE REVIEW

File: apps/next-evans/components/core/BookTeaserCard.tsx [hunk 1]
```diff
@@ -72,7 +76,7 @@ const BookTeaserCard: React.FC<Props> = (props) => {
             <Link href={props.documentUrl}>
               <Front
                 author={authorName}
-                edition={mostModernEdition(props.editionTypes)}
+                edition={mostModernEdition(editionTypes(props.editions))}
```
Comment (jaredh159) [inline apps/next-evans/components/core/BookTeaserCard.tsx:79]: hmmm... i'm only seeing the usages in this PR, maybe there are a bunch of others, but it seems like almost everywhere you call `mostModernEdition` you end up having an array of editions, and needing to extract the types using your `editionTypes` helper. would it be better to maybe just have `mostModernEdition` take an array of editions directly? maybe you could quick scan the rest of the usages and see if that is the overwhelming majority.

or, if there are a few important call sites where you just have an array of types, perhaps you could overload the function signature to accept _either_ and practice your TS skills.

File: apps/next-evans/components/core/BookTeaserCard.tsx [hunk 2]
```diff
@@ -11,8 +11,12 @@ import { LANG } from '@/lib/env';
 import { isCompilations } from '@/lib/friend';
 import Button from '@/components/core/Button';
 import { mostModernEdition } from '@/lib/editions';
+import { editionTypes } from '@/lib/document';
 
-export type Props = DocumentWithMeta & {
+export type Props = Omit<
```
Comment (jaredh159) [inline apps/next-evans/components/core/BookTeaserCard.tsx:16]: hmm... i know in the past, when picking between `Pick<T>` and `Omit<T>`, i've used the rule of thumb of just preferring whichever ended up being shorter.  

but i'm wondering if for these prop types, if maybe we should take a hard stance that we always use `Pick`. it's more explicit, doesn't accidentally randomly grow when we query new things from the db, and would reduce some diff churn, at the expense of in some places being a bit more verbose.  what do you think?

Comment (jaredh159) [reply apps/next-evans/components/core/BookTeaserCard.tsx:16]: but.. i wonder if we'd end up with suuuuper long types.  waffling.

File: apps/next-evans/components/pages/explore/RegionBlock.tsx [hunk 3]
```diff
@@ -5,7 +5,9 @@ import BookSlider from './BookSlider';
 import MapSlider from './MapSlider';
 
 interface Props {
-  books: (Omit<DocumentWithMeta, 'numPages' | 'size'> & { region: Region })[];
+  books: (Omit<DocumentWithMeta, 'numPages' | 'size' | 'featuredDescription'> & {
+    region: Region;
+  })[];
```
Comment (jaredh159) [inline apps/next-evans/components/pages/explore/RegionBlock.tsx:10]: can you convert this to use `Array<T>` instead of `T[]`? i'm finding this type a bit hard to parse mentally.

File: apps/next-evans/components/pages/home/FormatsBlock.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      </Dual.Frag>
+    </Heading>
+    <div className="hidden md:block BookMask--lg float-right h-[470px] w-[330px] [shape-outside:polygon(268px_99.28%,2.97%_1.75%,100%_0.25%,99.7%_99.57%)]"></div>
+    <div
+      className={cx(
+        `BookMask--sm float-right h-[700px] w-[130px] [shape-outside:polygon(173px_26.1%,59.89%_1.68%,126.14%_-5.03%,113.54%_103.04%,47.4%_81.24%,132.78%_73.29%,29.23%_37.77%)] md:hidden`,
+        `min-[400px]:h-[600px] min-[400px]:[shape-outside:polygon(167px_60.53%,27.58%_19.69%,119.22%_6.54%,107.39%_94.73%,46.15%_70.51%)]`,
+        `min-[500px]:h-[500px] min-[500px]:[shape-outside:polygon(165px_73.13%,16.04%_23.69%,127.68%_8.94%,108.16%_112.73%,43.07%_86.51%)]`,
+        `min-[600px]:h-[400px] min-[600px]:[shape-outside:polygon(169px_91.13%,127.58%_-0.56%,2.3%_23.01%)]`,
+        `min-[640px]:h-[450px] min-[640px]:[shape-outside:polygon(171px_74.69%,13.73%_12.36%,129.99%_-5.5%,95.85%_117.62%,33.07%_90.51%)]`,
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/FormatsBlock.tsx:27]: this almost makes me wonder if external css might be good sometimes.  not saying you should change it, just thinking aloud...

File: apps/next-evans/components/pages/home/FeaturedBooksBlock.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        setControlled(true);
+        setIndex(index === 0 ? books.length - 1 : index - 1);
+      }}
+      onSwipedLeft={() => {
+        setControlled(true);
+        setIndex(index === books.length - 1 ? 0 : index + 1);
+      }}
+    >
+      <Heading className="text-gray-800">{t`Featured Books`}</Heading>
+      <div className="BooksViewer w-[calc(100vw-(100vw-100%))] overflow-hidden">
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/FeaturedBooksBlock.tsx:41]: can you do a slow once-over of the changes in this PR looking for classes like `BooksViewer` that aren't tailwind utilities and checking if we can delete them? i'd like to avoid transferring classes that aren't doing anything to the new site if possible.

File: apps/next-evans/components/pages/home/SubHeroBlock.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import Heading from '@/components/core/Heading';
+import Dual from '@/components/core/Dual';
+import CoverImg from '@/public/images/samuel-fothergill-cover.jpg';
+import iPadImg from '@/public/images/ipad.png';
+import iPhoneImg from '@/public/images/iphone.png';
+import DeviceArrayImg from '@/public/images/device-array.png';
+import BookIcon from '@/public/images/icon-book.svg';
+import PhoneIcon from '@/public/images/icon-phone.svg';
+import HeadphonesIcon from '@/public/images/icon-headphones.svg';
+// import './SubHeroBlock.css';
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/SubHeroBlock.tsx:14]: whoops, delete.

File: apps/next-evans/components/pages/home/news-feed/news-feed.ts [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+}
+
+function getRecentAdditions(
+  lang: Lang,
+  documents: Array<NewsfeedDocumentProps>,
+  altLanguageDocuments: Array<NewsfeedDocumentProps> = [],
+): FeedItem[] {
+  const type = LANG === `en` && lang === `es` ? `spanish_translation` : `book`;
+  return documents
+    .sort((a, b) => new Date(b.dateAdded).getTime() - new Date(a.dateAdded).getTime())
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/news-feed/news-feed.ts:60]: what do you think about making some sort of lib util helper fn called `newestFirst` so we could clean up some of these noisy sorts to just read:

```tsx
  .sort((a, b) => newestFirst(a.date, b.date))
```

File: apps/next-evans/components/pages/home/news-feed/news-feed.ts [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+          lang === `en`
+            ? htmlShortTitle(doc.title)
+            : `${htmlShortTitle(doc.title)}${LANG === `en` ? ` &mdash; (Spanish)` : ``}`,
+        ...dateFields(doc.mostModernEdition.impressionCreatedAt, LANG),
+        description:
+          type === `book`
+            ? LANG === `en`
+              ? `Download free eBook or pdf, or purchase a paperback at cost.`
+              : `Descárgalo en formato ebook o pdf, o compra el libro impreso a precio de costo.`
+            : `${doc.authorName}'s <em>${
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/news-feed/news-feed.ts:74]: the apostrophe here should be a curly apostrophe. i'm a bit of a curmudgeon when it comes to the friends library sites, i try not to have any straight quotes in text.

File: apps/next-evans/components/pages/home/news-feed/news-feed.ts [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+                altLanguageDocuments.find((altDoc) => altDoc.altLanguageId === doc.id)
+                  ?.title ?? ``
+              }</em> is now translated and available on the Spanish site.`,
+        url:
+          type === `spanish_translation`
+            ? `https://bibliotecadelosamigos.org/${doc.authorSlug}/${doc.slug}`
+            : `/${doc.authorSlug}/${doc.slug}`,
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/news-feed/news-feed.ts:81]: i have a two-part proposal:

first, i think this line and the one above it should be calling out to our `getDocumentUrl()` helper function. not so much because it's shorter, as because it preserves a single source of truth. if we ever wanted to change how our urls are structured, it would be nice to know we could change that one function, and not miss any hand-constructed urls sprinkled thru the code base.

second, is i think it might be nice to overload that function so that it could take a `{authorSlug: string; slug: string }` arg as well as the two string args. then, it would be more convenient to use. if you had two strings, pass em; if you have a document-like object, pass it.

File: apps/next-evans/components/pages/home/news-feed/news-feed.ts [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+  >,
+): FeedItem[] {
+  return documents
+    .map((doc) =>
+      doc.editions.reduce<(Audiobook & { title: string; url: string }) | null>(
+        (acc, ed) => {
+          if (
+            ed.audiobook &&
+            new Date(ed.audiobook.dateAdded).getTime() >
+              new Date(acc ? acc.dateAdded : `January 1, 1970`).getTime()
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/news-feed/news-feed.ts:101]: could we shorten this to:

```ts
new Date(acc?.dateAdded  ??`January 1, 1970`).getTime()
```

File: apps/next-evans/components/pages/home/news-feed/news-feed.ts [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+          }
+          return acc;
+        },
+        null,
+      ),
+    )
+    .filter(isNotNull)
+    .map((audio) => ({
+      type: `audiobook`,
+      title: `${audio.title} &mdash; (${LANG === `en` ? `Audiobook` : `Audiolibro`})`,
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/news-feed/news-feed.ts:117]: let's use

```ts
t`Audiobook`
```

instead of the ternary. if you ever want to check if we have a translation for something, fuzzy find for `strings.ts` in the monorepo.

File: apps/next-evans/components/pages/home/news-feed/news-feed.ts [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+      url: audio.url,
+    }));
+}
+
+function dateFields(
+  dateStr: string,
+  lang: Lang,
+): Pick<FeedItem, 'month' | 'year' | 'day' | 'date'> {
+  const date = new Date(dateStr);
+  const month = months[lang][date.getMonth()]?.substring(0, 3) || ``;
```
Comment (jaredh159) [inline apps/next-evans/components/pages/home/news-feed/news-feed.ts:132]: because dates are weird and hard and a source of lots of bugs, i think i'd feel better about using `invariant` and getting a build-time error rather than falling back to the empty string here, b/c i think there's a non-trivial chance it might happen.

File: apps/next-evans/lib/db/friends.ts [hunk 13]
```diff
@@ -27,40 +25,108 @@ export async function getAllFriends(
...
+            created_at: null,
+            isbn: firstEdition.isbns[0]?.code ?? ``,
+            dateAdded: doc.created_at.toISOString(),
+            featuredDescription: doc.featured_description,
+            mostModernEdition: toEdition(newestEdition),
+            editions: doc.editions.map(toEdition),
+            shortDescription: doc.partial_description,
+            hasAudio: doc.editions.some((e) => e.edition_audios?.id),
+            tags: doc.document_tags.map((t) => t.type),
+            numDownloads: doc.editions.reduce((acc, ed) => acc + ed.downloads.length, 0), // TODO ~ maybe figure out if prisma can just count the downloads so don't have to get the ids
```
Comment (jaredh159) [inline apps/next-evans/lib/db/friends.ts:68]: can you take a minute and see if this TODO is possible/easy?

File: apps/next-evans/lib/db/friends.ts [hunk 14]
```diff
@@ -27,40 +25,108 @@ export async function getAllFriends(
...
+    return acc;
+  }, {});
 }
 
-export async function getNumDocuments(lang: Lang): Promise<number> {
-  return Object.values(await getAllDocuments(lang)).length;
+function toEdition(
+  apiEdition: Awaited<
+    ReturnType<typeof queryFriends>
+  >[number]['documents'][number]['editions'][number],
```
Comment (jaredh159) [inline apps/next-evans/lib/db/friends.ts:86]: i think it would be preferable to name this monstrous type outside the function definition. and i would propose maybe calling it `DbEdition` or  `QueryEdition` rather than "api-" because it's not actually coming from our api, it's just coming from the database. so the variable name `apiEdition` maybe should be changed too.

File: apps/next-evans/lib/db/friends.ts [hunk 15]
```diff
@@ -27,40 +25,108 @@ export async function getAllFriends(
...
+): Edition {
+  const audiobook = apiEdition.edition_audios;
+  invariant(apiEdition.edition_impressions !== null); // TODO ~ prisma types are wrong
```
Comment (jaredh159) [inline apps/next-evans/lib/db/friends.ts:89]: don't spend more than a few minutes on this, because i'm less confident of success, but can you also briefly see if there's a more correct way to query this that would give us the correct type inference?

File: apps/next-evans/lib/db/friends.ts [hunk 16]
```diff
@@ -104,19 +170,28 @@ async function getFriendsFromDB(lang: Lang): Promise<Record<string, FriendType>>
           created_at: true,
           slug: true,
           partial_description: true,
+          featured_description: true,
           id: true,
+          alt_language_id: true,
```
Comment (jaredh159) [inline apps/next-evans/lib/db/friends.ts:175]: i see you added a bunch of `featuredDescription`s to all of the `Omit`s, but does this line mean we should be omitting `altLanguageId` as well? it's this sort of thing that makes me wonder if we should go `Pick` only.

File: apps/next-evans/lib/types.ts [hunk 17]
```diff
@@ -42,7 +51,13 @@ export type DocumentWithMeta = DocumentType & {
   publishedDate: number | null;
 };
 
-export type Edition = 'original' | 'modernized' | 'updated';
+export interface Audiobook {
+  id: string;
+  isIncomplete: boolean;
+  dateAdded: string;
```
Comment (jaredh159) [inline apps/next-evans/lib/types.ts:57]: let's hold off for now, but i think some global typealiases for `UUID` and `IsoDateString` like in gertie, could be nice. i do fewer of these than i used to, but those two feel like they pull their weight to me.

File: apps/next-evans/pages/home.tsx [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+export const getStaticProps: GetStaticProps<Props> = async () => {
+  const documents = { en: await getAllDocuments(`en`), es: await getAllDocuments(`es`) };
+  let featuredBooks = [
+    documents.en[`compilations/truth-in-the-inward-parts-v1`],
+    documents.en[`hugh-turford/walk-in-the-spirit`],
+    documents.en[`isaac-penington/writings-volume-1`],
+    documents.en[`isaac-penington/writings-volume-2`],
+    documents.en[`william-penn/no-cross-no-crown`],
+    documents.en[`william-sewel/history-of-quakers`],
+  ].filter((doc): doc is DocumentWithMeta => doc !== undefined);
```
Comment (jaredh159) [inline apps/next-evans/pages/home.tsx:26]: if we changed the `notNull` helper to be called `notNullish` and had it reject `undefined` as well, we could use it here and on line 32 below.

**Review by jaredh159 — CHANGES_REQUESTED** [review]
solid work, this is really coming along nicely. left you quite a few proposed changes. 👍

File: apps/next-evans/lib/db/friends.ts (hunk 13 above)
Comment (kiahjh) [reply apps/next-evans/lib/db/friends.ts:68]: Just played around with it, I don't think you can...

Comment (kiahjh) [reply apps/next-evans/lib/db/friends.ts:68]: You can do it at the top level like this:
```
const count = await prisma.document.count();
```
But I don't think there's a way to do it way nested in there.

File: apps/next-evans/components/pages/explore/SearchBlock.tsx [hunk 19]
```diff
@@ -11,7 +11,29 @@ interface Props {
   initialFilters?: string[];
   initialUsed?: boolean;
   books: Array<
-    Omit<DocumentWithMeta, 'numPages' | 'size' | 'featuredDescription'> & {
+    Pick<
+      DocumentWithMeta,
+      //
```
Comment (jaredh159) [inline apps/next-evans/components/pages/explore/SearchBlock.tsx:16]: probably delete this line.

File: apps/next-evans/lib/friend.ts [hunk 20]
```diff
@@ -14,8 +14,16 @@ export function getFriendUrl(friendSlug: string, gender: Gender): string {
   return `/${firstPart}/${friendSlug}`;
 }
 
-export function getDocumentUrl(friendSlug: string, documentSlug: string): string {
-  return `/${friendSlug}/${documentSlug}`;
+export function getDocumentUrl(doc: { authorSlug: string; slug: string }): string;
+export function getDocumentUrl(str1: string, str2: string): string;
```
Comment (jaredh159) [inline apps/next-evans/lib/friend.ts:18]: this function signature will show up in our editors intellisense, so we want to preserve good names for variables, so these should still be `friendSlug` and `documentSlug`. the implementation can remain the same, it's effectively hidden to us at the call site.

**Review by jaredh159 — CHANGES_REQUESTED** [review]
super close, just a few small changes needed. :)

**Review by jaredh159 — APPROVED** [review]
